src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs

```diff
@@ -48,43 +46,74 @@ public async Task RestoreStateAsync(IPersistentComponentStateStore store)
     public Task PersistStateAsync(IPersistentComponentStateStore store, Renderer renderer)
-        => PersistStateAsync(store, renderer.Dispatcher);
-
-    /// <summary>
-    /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>.
-    /// </summary>
-    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to restore the application state from.</param>
-    /// <param name="dispatcher">The <see cref="Dispatcher"/> corresponding to the components' renderer.</param>
-    /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
-    public Task PersistStateAsync(IPersistentComponentStateStore store, Dispatcher dispatcher)
     {
-        if (_stateIsPersisted)
+        return renderer.Dispatcher.InvokeAsync(PauseAndPersistState);
+
+        async Task PauseAndPersistState()
         {
-            throw new InvalidOperationException("State already persisted.");
+            InferRenderModes(renderer);
+
+            if (store is IEnumerable<IPersistentComponentStateStore> compositeStore)
+            {
+                foreach (var store in compositeStore)
+                {
+                    await PersistState(store);
+                }
+            }
+            else
+            {
+                await PersistState(store);
+            }
         }
 
-        _stateIsPersisted = true;
+        async Task PersistState(IPersistentComponentStateStore store)
+        {
+            var currentState = new Dictionary<string, byte[]>();
+            State.PersistenceContext = new(currentState);
 
-        return dispatcher.InvokeAsync(PauseAndPersistState);
+            await PauseAsync(store);
+            await store.PersistStateAsync(currentState);
+        }
+    }
 
-        async Task PauseAndPersistState()
+    private void InferRenderModes(Renderer renderer)
+    {
+        for (var i = 0; i < _registeredCallbacks.Count; i++)
         {
-            State.PersistingState = true;
-            await PauseAsync();
-            State.PersistingState = false;
+            var registration = _registeredCallbacks[i];
+            if (registration.RenderMode != null)
+            {
+                // Explicitly set render mode, so nothing to do.
+                continue;
+            }
 
-            await store.PersistStateAsync(_currentState);
+            if (registration.Callback.Target is IComponent component)
+            {
+                var componentRenderMode = renderer.GetComponentRenderMode(component);
+                _registeredCallbacks[i] = new RegistrationContext(registration.Callback, componentRenderMode);
+                continue;
+            }
+
+            throw new InvalidOperationException(
+                $"The registered callback {registration.Callback.Method.Name} must be associated with a component or define" +
+                $" an explicit render mode type during registration.");
```

Comment:
Is this a breaking change?

Reply:
I'm pushing an update; it's not going to be a breaking change because we are only going to do this when we can't infer the render mode. In any other case, it won't be needed. There is a test that covers that behavior.